plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePartitionManager.java

```diff
@@ -98,27 +99,39 @@ public HivePartitionResult getPartitions(SemiTransactionalHiveMetastore metastor
         if (partitionColumns.isEmpty()) {
             return new HivePartitionResult(
                     partitionColumns,
+                    ImmutableList.of(UNPARTITIONED_ID),
                     ImmutableList.of(new HivePartition(tableName)),
                     compactEffectivePredicate,
                     effectivePredicate,
                     TupleDomain.all(),
                     hiveBucketHandle,
-                    bucketFilter);
+                    bucketFilter,
+                    true);
         }
 
         List<Type> partitionTypes = partitionColumns.stream()
                 .map(HiveColumnHandle::getType)
                 .collect(toList());
 
         Iterable<HivePartition> partitionsIterable;
+        List<String> partitionNames;
         Predicate<Map<ColumnHandle, NullableValue>> predicate = constraint.predicate().orElse(value -> true);
         if (hiveTableHandle.getPartitions().isPresent()) {
-            partitionsIterable = hiveTableHandle.getPartitions().get().stream()
+            List<HivePartition> partitions = hiveTableHandle.getPartitions().get().stream()
                     .filter(partition -> partitionMatches(partitionColumns, effectivePredicate, predicate, partition))
                     .collect(toImmutableList());
+            partitionsIterable = partitions;
+            partitionNames = partitions.stream()
+                    .map(HivePartition::getPartitionId)
+                    .collect(toImmutableList());
         }
         else {
-            List<String> partitionNames = getFilteredPartitionNames(metastore, identity, tableName, partitionColumns, compactEffectivePredicate);
+            if (hiveTableHandle.getPartitionNames().isPresent()) {
+                partitionNames = hiveTableHandle.getPartitionNames().get();
+            }
+            else {
+                partitionNames = getFilteredPartitionNames(metastore, identity, tableName, partitionColumns, compactEffectivePredicate);
+            }
             partitionsIterable = () -> partitionNames.stream()
```

Comment:
so `partitionNames` can be wider than `partitionsIterable` results in this branch (compared to `if (hiveTableHandle.getPartitions().isPresent()) {`)?

I think `HiveTableHandle` needs good Javadoc what is the contract of `partitionNames` and `partitions`.